kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license:
+//kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/openmp_utils.h"
+
+namespace Kratos {
+namespace Parallel {
+///@addtogroup Kratos
+///@{
+
+/// type definitions
+typedef int ThreadIdType;  // do not expect this to be an int. It is not for
+                           // C++11 parallelism
+
+inline int GetNumThreads() {
+#ifdef _OPENMP
+    return omp_get_max_threads();
+#else
+    return 1;
+#endif
+}
+
+inline ThreadIdType GetThreadHashId() {
+#ifdef _OPENMP
+    return omp_get_thread_num();
+#else
+    return 0;
+#endif
+}
+
+///@name Kratos Globals
+///@{
+// this function implements a parallel for, in which f is a lambda function
+// which captures its parameters once per very time it is called
+template <class InputIt, class UnaryFunction>
+inline void parallel_for(InputIt first, const int n, UnaryFunction &&f) {
+#pragma omp parallel for firstprivate(first)
+    for (int i = 0; i < n; ++i) {
+        f(first + i);
+    }
+}
+
+// parallel for executed by chunks. Here the function is called on every "block
+// of unknowns", so that the lambda captures are done just once
+// note that the function b must implement the loop from start to end
+// work allocation is static
+template <class TContainterType, class UnaryFunction>
+inline void parallel_for(TContainterType &data, UnaryFunction &&f) {
+    parallel_for(data.begin(), data.size(), f);
+}
+
+// the function f is executed in parallel. The parameter passed to f is the
+// number of the thread executing it.
+// one lambda capture is passed per every thread
+template <class UnaryFunction>
+inline void execute(UnaryFunction &&f) {
```

Comment:
The template might give strange compiler errors which are hard to understand if the user passes an incompatible function. We know here that we should take a function which accepts an int. Maybe it is better to do `execute(std::function<void(int)> f)`